impl/open_search.py

```diff
@@ -0,0 +1,226 @@
+from django.core.serializers import serialize
+from django.http import JsonResponse
+
+# import settings.settings
+from django.conf import settings
+from ezidapp.models.identifier import Identifier
+import json
+import pdb
+import base64
+import datetime
+import ezidapp.models.validation as validation
+import impl.util
+import requests
+from django.conf import settings
+from urllib.parse import quote
+import re
+import functools
+
+# the functools allows memoizing the results of functions so they're not recalculated every time (ie cached
+# results if called more than once on the same instance)
+
+MAX_SEARCHABLE_TARGET_LENGTH = 255
+INDEXED_PREFIX_LENGTH = 50
+
+# testing
+# python manage.py shell
+
+# seems like they're using the django db model libraries https://docs.djangoproject.com/en/5.0/topics/db/queries/
+
+"""
+-- basic testing
+import impl.open_search as os
+from ezidapp.models.identifier import Identifier
+open_s = os.OpenSearch(identifier=Identifier.objects.get(identifier='doi:10.25338/B8JG7X'))
+my_dict = open_s.dict_for_identifier()
+"""
+
+# todo: Do we need more meaningful values for these fields or are the database IDs ok?
+# datacenter, profile, owner, ownergroup
+# also, do we really need the crossref status and message which seem like internal fields for EZID maintenance, not search
+class OpenSearch:
+    def __init__(self, identifier: Identifier):
+        self.identifier = identifier
+        self.km = identifier.kernelMetadata
+
+    # someone broke Python conventions and wrote some fields as camelCase instead of snake_case, so don't want to
+    # propagate it further
+    def _camel_to_snake(name):
+        name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+        return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()
+
+    def dict_for_identifier(self) -> str:
+        identifier_dict = {}
+        exclude_fields = "pk cm metadata".split()
+        for field in self.identifier._meta.fields:
+            if field.name not in exclude_fields:
+                field_name = OpenSearch._camel_to_snake(field.name)
+                field_value = getattr(self.identifier, field.name)
+                if field_value is None:
+                    field_value = ''
+
+                # this fugly construct because . . . python
+                if hasattr(field_value, 'pk'):  # Check if it's a foreign key
```

Comment:
We may want to retrieve the mapped value from the foreign key.

Reply:
I'm including the primary key as db_identifier_id in the json.